File: src/main/java/org/micoli/php/configuration/schema/valueGenerator/IconValueGenerator.java
package org.micoli.php.configuration.schema.valueGenerator;

import java.io.IOException;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.*;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;
import org.jetbrains.annotations.NotNull;

public class IconValueGenerator implements PropertyValueGenerator {
    List<String> fieldNames = List.of("icon", "activeIcon", "inactiveIcon", "unknownIcon");

    @Override
    public ReferenceType getType() {
        return ReferenceType.AS_REF;
    }

    @Override
    public String getRefId() {
        return "icons";
    }

    @Override
    public List<String> getFieldNames() {
        return fieldNames;
    }

    @Override
    public List<String> getValues() {
        return findAllExpUISVGResources();
    }

    public List<String> findAllExpUISVGResources() {
        List<String> resources = new ArrayList<>();
        final String EXP_UI_PATH = "expui";
        try {
            URL url = getClass().getClassLoader().getResource(EXP_UI_PATH);
            if (url == null) {
                return resources;
            }

            String protocol = url.getProtocol();
            switch (protocol) {
                case "file" -> {
                    Path basePath = Paths.get(url.toURI());
                    Files.walkFileTree(basePath, new SimpleFileVisitor<>() {
                        @Override
                        public @NotNull FileVisitResult visitFile(
                                @NotNull Path file, @NotNull BasicFileAttributes attrs) {
                            String filePath = file.toString();
                            if (!attrs.isDirectory()
                                    && filePath.startsWith(EXP_UI_PATH + "/")
                                    && filePath.endsWith(".svg")) {
                                resources.add(filePath);
                            }
                            return FileVisitResult.CONTINUE;
                        }
                    });
                }
                case "jar" -> {
                    String jarPath = url.getPath()
                            .substring(EXP_UI_PATH.length(), url.getPath().indexOf("!"));
                    try (JarFile jar = new JarFile(URLDecoder.decode(jarPath, StandardCharsets.UTF_8))) {
                        Enumeration<JarEntry> entries = jar.entries();
                        while (entries.hasMoreElements()) {
                            JarEntry entry = entries.nextElement();
                            String filePath = entry.getName();
                            if (!entry.isDirectory()
                                    && filePath.startsWith(EXP_UI_PATH + "/")
                                    && filePath.endsWith(".svg")) {
                                resources.add(filePath);
                            }
                        }
                    } catch (IOException ignored) {
                    }
                }
            }
        } catch (URISyntaxException | IOException ignored) {
        }

        return resources.stream().filter(Objects::nonNull).distinct().sorted().toList();
    }
}

File: src/main/kotlin/org/micoli/php/configuration/schema/valueGenerator/IconValueGenerator.kt
package org.micoli.php.configuration.schema.valueGenerator

import java.io.IOException
import java.net.URISyntaxException
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.nio.file.*
import java.nio.file.attribute.BasicFileAttributes
import java.util.*
import java.util.jar.JarFile
import kotlinx.collections.immutable.ImmutableList
import kotlinx.collections.immutable.persistentListOf
import kotlinx.collections.immutable.toImmutableList

class IconValueGenerator : PropertyValueGenerator {
    override fun getFieldNames(): ImmutableList<String> = persistentListOf("icon", "activeIcon", "inactiveIcon", "unknownIcon")

    override fun getValues(): ImmutableList<String> {
        val expUIPath = "expui"
        val resources: MutableList<String> = ArrayList<String>()
        try {
            val url = javaClass.getClassLoader().getResource(expUIPath) ?: return resources.toImmutableList()

            val protocol = url.protocol
            when (protocol) {
                "file" -> {
                    val basePath = Paths.get(url.toURI())
                    Files.walkFileTree(
                      basePath,
                      object : SimpleFileVisitor<Path>() {
                          override fun visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult {
                              val filePath = file.toString()
                              if (!attrs.isDirectory && filePath.startsWith("$expUIPath/") && filePath.endsWith(".svg")) {
                                  resources.add(filePath)
                              }
                              return FileVisitResult.CONTINUE
                          }
                      },
                    )
                }
                "jar" -> {
                    val jarPath = url.path.substring(expUIPath.length, url.path.indexOf("!"))
                    try {
                        JarFile(URLDecoder.decode(jarPath, StandardCharsets.UTF_8)).use { jar ->
                            val entries = jar.entries()
                            while (entries.hasMoreElements()) {
                                val entry = entries.nextElement()
                                val filePath = entry.getName()
                                if (!entry.isDirectory && filePath.startsWith("$expUIPath/") && filePath.endsWith(".svg")) {
                                    resources.add(filePath)
                                }
                            }
                        }
                    } catch (_: IOException) {}
                }
            }
        } catch (_: URISyntaxException) {} catch (_: IOException) {}

        return resources.stream().filter { obj: String? -> Objects.nonNull(obj) }.distinct().sorted().toList().toImmutableList()
    }
}
